berkeley-eecs/seating
# Refac: More Modularization with Better Dir Structure

Improve the folder structure a bit by modularizing stuff as much as possible. 

Now, here is what it looks like in full:

```
/seating
    /server
        /static ------------------------- stylesheets, images
        /templates -------------------- jinja2 templates (views)
        /controllers ------------------- modularized controllers
            __init__.py
            auth_controllers.py -------- login, logout, callback routes etc.
            health_controllers.py ------ basic health check routes
        /services ---------------------- micro services, business logic
			__init__.py
			/auth ----------------------- auth clients etc
			/core ----------------------- data imports, assignment algo etc.
			/canvas --------------------- canvas client
			/email ----------------------  all email related features (tbd)
		/typings ----------------------- enums, types
        models.py --------------------- models
        forms.py ----------------------- wtforms 
        views.py ----------------------- the other controllers ... to be modularized
		__init__.py ---------------------  app entry point
    /tests
        __init__.py
        conftest.py -------------------- pytest config, fixtures, scaffolding
        /unit --------------------------- contains unit tests
		/e2e --------------------------- contains end-to-end tests (ui tests)
    /cli
		__init__.py
		db.py --------------------------- flask cli commands related to DB
		qa.py --------------------------- flask cli commands related to QA
    config.py -------------------------- flask env-specific config
	.env.sample
    requirements.txt
	requirements-dev.txt
	---------  the following are tooling configs
	.flake8
	/.github
	/.vscode
	README.md
	Procfile ---------------------------- not sure what this does, prob for deployment
	runtime.txt ------------------------ not sure what this does, prob for deployment
```

A few highlights for this PR:
- Create `server/controllers/` to hold route handlers.

Currently, only contains auth and health controllers. Should gradually break up and move controllers from the big `server/view.py` file to this.
Due to the use of [url convertors](https://exploreflask.com/en/latest/views.html) this might get complicated.

Also, I don't get why everyone insists on having a `views.py` at application root which holds controllers (and, even broken down to different controller folders, people still name the controller file as`views`). You could call it controllers/ handlers / router_handlers or even routers but like why views?

- Create `server/services` to hold micro services within this app.

Originally we have `util` folder where I dumped in everything that are not MVC. A misnomer. I deleted the util and moved its files around and the majority are kept in this services folder. The idea is to go for a **Domain-driven design** or Multiple Microservices kind of architecture, where each service deals with its own business and talks to each other with their public API. We have an email service, canvasapi service, auth services etc.

It is a little challenging to set up as I find it hard in Python to draw a boundary between private and public. The abstraction layer is only loosely kept. The convention I use is roughly that other parts of the code should only import the `__init__.py` of each service.

- Create `cli` to hold those cli commands that are scattered around.

As title.
